### PR TITLE
Replace max tensor with max cut

### PR DESCRIPTION
**Pull request by elad-c**
## Pull Request Description:
Replace max tensor with max cut in resource utilization. 

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/common/graph/memory_graph/memory_graph.py [hunk 1]
```diff
@@ -45,7 +45,11 @@ def __init__(self, model_graph: Graph):
         tensor_to_node = []
 
         for n in nodes:
-            n_outputs = [n.output_shape] if isinstance(n.output_shape, tuple) else n.output_shape
+            if len(n.output_shape) == 0:
+                n_outputs = []
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/graph/memory_graph/memory_graph.py:49]: just want to verify - what case is this handling? what is n?

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py [hunk 2]
```diff
@@ -66,7 +67,10 @@ def __init__(self,
         self.layer_to_bitwidth_mapping = self.get_search_space()
         self.compute_metric_fn = self.get_sensitivity_metric()
 
-        self.compute_ru_functions = ru_functions
+        self.cuts = calc_graph_cuts(self.original_graph)
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py:70]: Maybe add a todo to find a better solution for running the cuts computation? 
The goal is that we won't need to run the max cut computation in case we are not in activation (or total) constrained mixed precision.

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py [hunk 3]
```diff
@@ -66,7 +67,10 @@ def __init__(self,
...
+
+        ru_types = [k for k, v in target_resource_utilization.get_resource_utilization_dict().items() if v < np.inf]
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py:72]: change k, v to "ru_target, ru_value" or something like that, so it would be easier to understand what to expect in this list

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py [hunk 4]
```diff
@@ -66,7 +67,10 @@ def __init__(self,
...
+        self.compute_ru_functions = {k: v for k, v in ru_functions.items() if k in ru_types}
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py:73]: same (k, v names)

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py [hunk 5]
```diff
@@ -121,7 +125,10 @@ def compute_min_ru(self) -> Dict[RUTarget, np.ndarray]:
         for ru_target, ru_fns in self.compute_ru_functions.items():
             # ru_fns is a pair of resource utilization computation method and 
             # resource utilization aggregation method (in this method we only need the first one)
-            min_ru[ru_target] = ru_fns[0](self.min_ru_config, self.graph, self.fw_info, self.fw_impl)
+            if ru_target is RUTarget.ACTIVATION:
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py:128]: is "is" correct? usually enum is compared with '==' (I think it is equivalent for enums but just want to make sure)

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py [hunk 6]
```diff
@@ -212,7 +219,10 @@ def compute_node_ru_for_candidate(self, conf_node_idx: int, candidate_idx: int,
 
         """
         cfg = self.replace_config_in_index(self.min_ru_config, conf_node_idx, candidate_idx)
-        return self.compute_ru_functions[target].metric_fn(cfg, self.graph, self.fw_info, self.fw_impl)
+        if target == RUTarget.ACTIVATION:
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py:222]: it's duplicated with the code under  compute_min_ru just the cfg is different. maybe extract this to a class protected function?

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py [hunk 7]
```diff
@@ -87,6 +90,88 @@ def weights_size_utilization(mp_cfg: List[int],
...
+    """
+    Calculate graph activation cuts.
+    Args:
+        graph: A graph object to calculate activation cuts on.
+
+    Returns:
+        A list of activation cuts.
+
+    """
+    memory_graph = MemoryGraph(deepcopy(graph))  # TODO maxcut: do we need the deepcopy?
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py:103]: do we need the deepcopy? (:

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py [hunk 8]
```diff
@@ -87,6 +90,88 @@ def weights_size_utilization(mp_cfg: List[int],
...
+    _, _, cuts = compute_graph_max_cut(memory_graph)
+
+    if cuts is None:
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py:106]: `
filtered_cuts = None 
if cuts is not None: 
... 

return filtered_cuts
`



I think it's simpler but up to you

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py [hunk 9]
```diff
@@ -87,6 +90,88 @@ def weights_size_utilization(mp_cfg: List[int],
...
+        return None
+    else:
+        # filter empty cuts and cuta that contain nodes with activation quantization disabled.
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py:109]: cuta --> cuts ?

Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py:109]: any node with activation quantization disabled?
didn't we agree that we should not remove those cuts but just remove the memory elements from the computation?

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py [hunk 10]
```diff
@@ -87,6 +90,88 @@ def weights_size_utilization(mp_cfg: List[int],
...
+        filtered_cuts = []
+        for cut in cuts:
+            if len(cut.mem_elements.elements) > 0 and any(
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py:112]: consider assgining the condition to a variable and make the if condition clearer

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py [hunk 11]
```diff
@@ -87,6 +90,88 @@ def weights_size_utilization(mp_cfg: List[int],
...
+                     for e in cut.mem_elements.elements]):
+                filtered_cuts.append(cut)
+        return filtered_cuts
+
+
+def activation_maxcut_size_utilization(mp_cfg: List[int],
+                                       graph: Graph,
+                                       fw_info: FrameworkInfo,
+                                       fw_impl: FrameworkImplementation,
+                                       cuts: List[Cut] = None) -> np.ndarray:
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py:123]: use Optional type

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py [hunk 12]
```diff
@@ -87,6 +90,88 @@ def weights_size_utilization(mp_cfg: List[int],
...
+            (not used in this method).
+        fw_impl: FrameworkImplementation object with specific framework methods implementation(not used in this method).
+        cuts: a list of graph cuts (optional. if not provided calculated locally).
+
+    Returns: A vector of node's cut memory sizes.
+    Note that the vector is not necessarily of the same length as the given config.
+
+    """
+    activation_cut_memory = []
+    if len(mp_cfg) == 0:
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py:141]: suggestion for minor change:
- Change the comment for an informative one that states that non-configurable nodes are not included in the cuts memory.
- init activation_cut_memory  with a cut of size 0
- remove 'if'
- add an assertion that node_idx < len(mp_cfg)

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py [hunk 13]
```diff
@@ -433,7 +518,7 @@ class MpRuMetric(Enum):
     """
 
     WEIGHTS_SIZE = partial(weights_size_utilization)
-    ACTIVATION_OUTPUT_SIZE = partial(activation_output_size_utilization)
+    ACTIVATION_OUTPUT_SIZE = partial(activation_maxcut_size_utilization)
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py:521]: why replace? why not add a new value to the enum of "ACTIVATION_CUTS_SIZE", and in the ru dict that is used for MP replace the activation target method to max cut?

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py [hunk 14]
```diff
@@ -132,6 +133,54 @@ def compute_nodes_weights_params(graph: Graph, fw_info: FrameworkInfo) -> Tuple[
 
     return np.array(weights_memory_bytes), np.array(weights_params)
 
+
+def compute_activation_output_maxcut_sizes(graph: Graph) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py:137]: This is eventually supposed to be removed and replaced with a single implementation of the ru metrics.
so no need to further optimize this code,
but @irenaby please take a look to make sure that this can be integrated into your change to ru data

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py [hunk 15]
```diff
@@ -238,7 +285,7 @@ def requires_mixed_precision(in_model: Any,
     total_weights_memory_bytes = 0 if len(weights_memory_by_layer_bytes) == 0 else sum(weights_memory_by_layer_bytes)
 
     # Compute max activation tensor in bytes
-    activation_output_sizes_bytes, _ = compute_activation_output_sizes(transformed_graph)
+    activation_output_sizes_bytes, _ = compute_activation_output_maxcut_sizes(transformed_graph)
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py:288]: change the variable name (maybe to generic "activation_memory_estimation_bytes" or something like that)

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/scaled_dot_product_attention.py [hunk 16]
```diff
@@ -68,8 +68,8 @@ def _get_transpose_k_node(self, attention_node_name: str, key_node: BaseNode) ->
         output_shape[-2], output_shape[-1] = input_shape[-1], input_shape[-2]
         transpose_node = FunctionalNode(name=f"{attention_node_name}_{key_node.name}_transpose",
                                         framework_attr={},
-                                        input_shape=input_shape,
-                                        output_shape=output_shape,
+                                        input_shape=[input_shape],
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/scaled_dot_product_attention.py:71]: is this change related to the max cut integration? if not, why is it failing all of a sudden? (similar question to the other changes in this file)

File: tests/keras_tests/feature_networks_tests/feature_networks/activation_16bit_test.py [hunk 17]
```diff
@@ -98,7 +98,7 @@ def create_networks(self):
 
     def compare(self, quantized_model, float_model, input_x=None, quantization_info=None):
         mul1_act_quant = quantized_model.layers[3]
-        mul2_act_quant = quantized_model.layers[9]
+        mul2_act_quant = quantized_model.layers[10]
```
Comment (ofirgo) [inline tests/keras_tests/feature_networks_tests/feature_networks/activation_16bit_test.py:101]: we should use the util function get_layers_from_model_by_type if I'm not mistaken (should eliminate the need for doing such modification for tests when our inner representation is changing)

File: tests/pytorch_tests/model_tests/test_feature_models_runner.py [hunk 18]
```diff
@@ -605,10 +605,10 @@ def test_mixed_precision_bops_utilization(self):
         MixedPrecisionBopsAllWeightsLayersTest(self).run_test()
         MixedPrecisionWeightsOnlyBopsTest(self).run_test()
         MixedPrecisionActivationOnlyBopsTest(self).run_test()
-        MixedPrecisionBopsAndWeightsMemoryUtilizationTest(self).run_test()
-        MixedPrecisionBopsAndActivationMemoryUtilizationTest(self).run_test()
-        MixedPrecisionBopsAndTotalMemoryUtilizationTest(self).run_test()
-        MixedPrecisionBopsWeightsActivationUtilizationTest(self).run_test()
+        # MixedPrecisionBopsAndWeightsMemoryUtilizationTest(self).run_test()
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/test_feature_models_runner.py:608]: add a todo regarding bringing back these tests once we solve the issue with combined bops with other ru metrics

File: tests/pytorch_tests/model_tests/feature_models/mixed_precision_activation_test.py [hunk 19]
```diff
@@ -112,7 +112,7 @@ def compare(self, quantized_models, float_model, input_x=None, quantization_info
 class MixedPrecisionActivationSearch4BitFunctional(MixedPrecisionActivationBaseTest):
     def __init__(self, unit_test):
         super().__init__(unit_test)
-        self.expected_config = [1, 4, 4, 1]
+        self.expected_config = [2, 5, 5, 1]
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/feature_models/mixed_precision_activation_test.py:115]: did you verify that this result make sense? (as far as possible, I mean by looking at the max cut result and observing the network structure)?

File: tests/pytorch_tests/model_tests/feature_models/mixed_precision_activation_test.py [hunk 20]
```diff
@@ -127,7 +127,7 @@ def compare(self, quantized_models, float_model, input_x=None, quantization_info
 class MixedPrecisionActivationMultipleInputs(MixedPrecisionActivationBaseTest):
     def __init__(self, unit_test):
         super().__init__(unit_test)
-        self.expected_config = [0 for _ in range(8)] + [1] # expected config for this test.
+        self.expected_config = [0, 0, 0, 0, 0, 0, 1, 0, 1]  # expected config for this test.
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/feature_models/mixed_precision_activation_test.py:130]: same here

File: tests/pytorch_tests/model_tests/feature_models/manual_bit_selection.py [hunk 21]
```diff
@@ -218,8 +219,7 @@ def get_tpc(self):
         return {'mixed_precision_activation_model': tpc}
 
     def get_resource_utilization(self):
-        return mct.core.ResourceUtilization(activation_memory=6200)
-
+        return mct.core.ResourceUtilization(activation_memory=15000)
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/feature_models/manual_bit_selection.py:223]: why this change? is it just to prevent auto mp execution?

File: tests/keras_tests/feature_networks_tests/test_features_runner.py [hunk 22]
```diff
@@ -322,10 +322,10 @@ def test_mixed_precision_bops_utilization(self):
         MixedPrecisionBopsAllWeightsLayersTest(self).run_test()
         MixedPrecisionWeightsOnlyBopsTest(self).run_test()
         MixedPrecisionActivationOnlyBopsTest(self).run_test()
-        MixedPrecisionBopsAndWeightsUtilizationTest(self).run_test()
-        MixedPrecisionBopsAndActivationUtilizationTest(self).run_test()
-        MixedPrecisionBopsAndTotalUtilizationTest(self).run_test()
-        MixedPrecisionBopsWeightsActivationUtilizationTest(self).run_test()
+        # MixedPrecisionBopsAndWeightsUtilizationTest(self).run_test()
```
Comment (ofirgo) [inline tests/keras_tests/feature_networks_tests/test_features_runner.py:325]: todo

**Review by ofirgo — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py (hunk 8 above)
Comment (irenaby) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py:106]: When can it be None?

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py [hunk 23]
```diff
@@ -87,6 +90,88 @@ def weights_size_utilization(mp_cfg: List[int],
...
+    Returns:
+        A list of activation cuts.
+
+    """
+    memory_graph = MemoryGraph(deepcopy(graph))  # TODO maxcut: do we need the deepcopy?
+    _, _, cuts = compute_graph_max_cut(memory_graph)
+
+    if cuts is None:
+        return None
+    else:
```
Comment (irenaby) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/ru_methods.py:108]: you don't need else

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py [hunk 24]
```diff
@@ -132,6 +133,54 @@ def compute_nodes_weights_params(graph: Graph, fw_info: FrameworkInfo) -> Tuple[
...
+    """
+    cuts = calc_graph_cuts(graph)
+
+    # map nodes to cuts.
+    node_to_cat_mapping = {}
+    for i, cut in enumerate(cuts):
+        mem_element_names = [m.node_name for m in cut.mem_elements.elements]
+        for m_name in mem_element_names:
+            if len(graph.find_node_by_name(m_name)) > 0:
+                if m_name in node_to_cat_mapping:
```
Comment (irenaby) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py:160]: from collections import defaultdict
node_to_cat_mapping = defaultdict(list))
node_to_cat_mapping[m_name].append(i)

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py [hunk 25]
```diff
@@ -132,6 +133,54 @@ def compute_nodes_weights_params(graph: Graph, fw_info: FrameworkInfo) -> Tuple[
...
+                    node_to_cat_mapping[m_name].append(i)
+                else:
+                    node_to_cat_mapping[m_name] = [i]
+            else:
+                raise Exception("Missing node")
+
+    activation_outputs = np.zeros(len(cuts))
+    activation_outputs_bytes = np.zeros(len(cuts))
+    for n in graph.nodes:
+        # Go over all nodes that have configurable activation.
```
Comment (irenaby) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py:170]: The comment says configurable (i.e. >1 q candidate), the code says any quantizable

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py [hunk 26]
```diff
@@ -132,6 +133,54 @@ def compute_nodes_weights_params(graph: Graph, fw_info: FrameworkInfo) -> Tuple[
...
+        if n.has_activation_quantization_enabled_candidate():
```
Comment (irenaby) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py:171]: It will be more readable to either filter the nodes before iterating, or
if not n.has...:
    continue 
.....

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py [hunk 27]
```diff
@@ -132,6 +133,54 @@ def compute_nodes_weights_params(graph: Graph, fw_info: FrameworkInfo) -> Tuple[
...
+            # Fetch maximum bits required for activations quantization.
+            max_activation_bits = max([qc.activation_quantization_cfg.activation_n_bits for qc in n.candidates_quantization_cfg])
+            node_output_size = n.get_total_output_params()  # TODO maxcut: use memory elements
```
Comment (irenaby) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py:174]: What's with TODO?

File: model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py [hunk 28]
```diff
@@ -132,6 +133,54 @@ def compute_nodes_weights_params(graph: Graph, fw_info: FrameworkInfo) -> Tuple[
...
+    for i, cut in enumerate(cuts):
+        mem_element_names = [m.node_name for m in cut.mem_elements.elements]
+        for m_name in mem_element_names:
+            if len(graph.find_node_by_name(m_name)) > 0:
+                if m_name in node_to_cat_mapping:
+                    node_to_cat_mapping[m_name].append(i)
+                else:
+                    node_to_cat_mapping[m_name] = [i]
+            else:
+                raise Exception("Missing node")
```
Comment (irenaby) [inline model_compression_toolkit/core/common/mixed_precision/resource_utilization_tools/resource_utilization_data.py:165]: generic exception is considered a bad practice

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py (hunk 5 above)
Comment (elad-c) [reply model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py:128]: it's the same. "is" feels clearer

File: tests/pytorch_tests/model_tests/feature_models/manual_bit_selection.py (hunk 21 above)
Comment (elad-c) [reply tests/pytorch_tests/model_tests/feature_models/manual_bit_selection.py:223]: It's larger because max cut is larger than max tensor.

File: model_compression_toolkit/core/common/graph/memory_graph/memory_graph.py [hunk 29]
```diff
@@ -71,6 +82,7 @@ def __init__(self, model_graph: Graph):
         inputs_tensors_memory = [sum([t.total_size for t in self.operation_node_children(n)])
                                  for n in nodes if n in model_graph.get_inputs()]
 
+        # TODO maxcut: why both inputs and outputs of each nodes, while the A* solves for node outputs only???
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/graph/memory_graph/memory_graph.py:85]: We need to deeply get into the algorithm to understand this.
If it is not giving you trouble then leave it for now.
Or, create some unittest that you for sure know what the result should be and see if changing this code according to what you think it should be changes the result.

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py [hunk 30]
```diff
@@ -76,6 +80,17 @@ def __init__(self,
         self.config_reconstruction_helper = ConfigReconstructionHelper(virtual_graph=self.graph,
                                                                        original_graph=self.original_graph)
 
+    @property
+    def cuts(self):
+        """
+        Calcualtes graph cuts. Written as property so it will only be calculkated once and
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py:86]: this is the way to do cacheing? with property? (maybe consult Irena if this is the best way to do this)
also a typo

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py [hunk 31]
```diff
@@ -76,6 +80,17 @@ def __init__(self,
         self.config_reconstruction_helper = ConfigReconstructionHelper(virtual_graph=self.graph,
                                                                        original_graph=self.original_graph)
 
+    @property
+    def cuts(self):
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py:84]: return typehint

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py [hunk 32]
```diff
@@ -106,6 +121,21 @@ def get_sensitivity_metric(self) -> Callable:
 
         return self.sensitivity_evaluator.compute_metric
 
+    def _calc_ru_fn(self, ru_target, ru_fn, mp_cfg):
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py:124]: return typehint

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py [hunk 33]
```diff
@@ -106,6 +121,21 @@ def get_sensitivity_metric(self) -> Callable:
...
+        """
+        Computes a resource utilization for a certain mp configuration
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py:126]: change mp to mixed precision in documentation

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py [hunk 34]
```diff
@@ -106,6 +121,21 @@ def get_sensitivity_metric(self) -> Callable:
...
+        The method computes a resource utilization vector for specific target resource utilization.
+
+        Returns: resource utilization value.
+
+        """
+        # ru_fn is a pair of resource utilization computation method and
+        # resource utilization aggregation method (in this method we only need the first one)
+        if ru_target is RUTarget.ACTIVATION:
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py:134]: maybe add a todo to eliminate this condition. this requires a refactor in the search manager but we need to include this in the refactor

File: tests/pytorch_tests/model_tests/feature_models/mixed_precision_activation_test.py [hunk 35]
```diff
@@ -127,7 +128,8 @@ def compare(self, quantized_models, float_model, input_x=None, quantization_info
 class MixedPrecisionActivationMultipleInputs(MixedPrecisionActivationBaseTest):
     def __init__(self, unit_test):
         super().__init__(unit_test)
-        self.expected_config = [0 for _ in range(8)] + [1] # expected config for this test.
+        # TODO maxcut: verify expected_config change is reasonable (was all zeros)
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/feature_models/mixed_precision_activation_test.py:131]: did you verify this?

File: tests/pytorch_tests/model_tests/feature_models/mixed_precision_activation_test.py [hunk 36]
```diff
@@ -112,7 +112,8 @@ def compare(self, quantized_models, float_model, input_x=None, quantization_info
 class MixedPrecisionActivationSearch4BitFunctional(MixedPrecisionActivationBaseTest):
     def __init__(self, unit_test):
         super().__init__(unit_test)
-        self.expected_config = [1, 4, 4, 1]
+        # TODO maxcut: verify expected_config change is reasonable (was [1, 4, 4, 1])
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/feature_models/mixed_precision_activation_test.py:115]: did you verify this?

File: tests/pytorch_tests/function_tests/resource_utilization_data_test.py [hunk 37]
```diff
@@ -172,7 +173,8 @@ def run_test(self):
 
         ru_data = prep_test(model, mp_bitwidth_candidates_list, large_random_datagen)
 
-        self.verify_results(ru_data, sum_parameters, max_tensor)
+        #  TODO maxcut: change to max cut. debug why max cut isn't 168003 (conv output + size). Currently fails periodically.
```
Comment (ofirgo) [inline tests/pytorch_tests/function_tests/resource_utilization_data_test.py:176]: did you verify this?

File: tests/pytorch_tests/function_tests/resource_utilization_data_test.py [hunk 38]
```diff
@@ -186,4 +188,5 @@ def run_test(self):
...
```
Comment (ofirgo) [inline tests/pytorch_tests/function_tests/resource_utilization_data_test.py:191]: did you verify this?

**Review by ofirgo — APPROVED** [review]
Mostly comments regarding TODOs that can be handled in the next PR.

Two other things:
- Most of the "TODOs" for non-cover code lines need to be covered
- We need to remember to insert an "experimental" message (or even as part of the RU target API?) for using the mp with max cut feature (also in the next PR)

**Review by irenaby — APPROVED** [review]

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py (hunk 34 above)
Comment (elad-c) [reply model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py:134]: This appears in multiple places. I'll add the TODO in the function itself.

File: tests/pytorch_tests/model_tests/feature_models/mixed_precision_activation_test.py (hunk 35 above)
Comment (elad-c) [reply tests/pytorch_tests/model_tests/feature_models/mixed_precision_activation_test.py:131]: in the next PR

File: tests/pytorch_tests/function_tests/resource_utilization_data_test.py (hunk 37 above)
Comment (elad-c) [reply tests/pytorch_tests/function_tests/resource_utilization_data_test.py:176]: No, I'll debug this in the next PR.

File: tests/pytorch_tests/function_tests/resource_utilization_data_test.py (hunk 38 above)
Comment (elad-c) [reply tests/pytorch_tests/function_tests/resource_utilization_data_test.py:191]: next PR
